New Kubernetes Superpowers blog post (#3925)
pulumi/docs: content/blog/new-kubernetes-superpowers/index.md

---
title: "New Kubernetes Superpowers for Cloud Native Infrastructure"
date: 2020-08-12

> meta_desc: "Announcing new features that advance Pulumi's Kubernetes support for deployment automation, cloud native ecosystem integration, and easy Puulumi adoption."

Reviewer: ```suggestion
meta_desc: "Announcing new features that advance Pulumi's Kubernetes support for deployment automation, cloud native ecosystem integration, and easy Pulumi adoption."
```

Whether you are designing your cloud native delivery platform, looking to enforce policy across cloud infrastructure and Kubernetes, or looking to bring robust software engineering practices to your Kubernetes workloads, these features continue to make Pulumi [a better way to Kubernetes](https://www.pulumi.com/blog/pulumi-a-better-way-to-kubernetes/)!

## Pulumi + Kubernetes

Pulumi is an Infrastructure as Code platform designed from the ground up for the modern cloud. That means deep support for containers, serverless, and Kubernetes, using your favorite programming languages to define, test, and manage your cloud infrastructure. Pulumi supports easy provisioning of Kubernetes clusters in many different managed platforms, including AWS, Azure, GCP, Digital Ocean, Rancher, Linode, and Civo. Additionally, Pulumi can manage resources inside a Kubernetes cluster using the same powerful tools, replacing the need for brittle and unscalable YAML configuration. Best of all, Pulumi combines these into a single cohesive programming model across your cloud infrastructure in managed cloud services *and* in Kubernetes.

> Kubernetes has been at the center of many Pulumi-enabled cloud modernization projects, including provisioning managed Kubernetes clusters, connecting them to supporting cloud infrastructure,  enabling platform teams to configure and manage platforms on top of Kubernetes, and allowing teams to replace unmaintainable piles of YAML with code and software-based deployment processes. Hundreds of modern cloud engineering teams have been able to accelerate the move to the cloud and Kubernetes thanks to Pulumi.

Reviewer: ```suggestion
Kubernetes has been at the center of many Pulumi-enabled cloud modernization projects, including provisioning managed Kubernetes clusters, connecting them to supporting cloud infrastructure, enabling platform teams to configure and manage platforms on top of Kubernetes, and allowing teams to replace unmaintainable piles of YAML with code and software-based deployment processes. Hundreds of modern cloud engineering teams have been able to accelerate the move to the cloud and Kubernetes thanks to Pulumi.
```

## Pulumi Kubernetes Operator

> Pulumi integrates into [more than a dozen CI/CD and delivery systems](https://www.pulumi.com/docs/guides/continuous-delivery/), automating the delivery of cloud infrastructure changes to AWS, Azure, GCP, Kubernetes, and more. This includes traditional CI tools like [Jenkins](https://www.pulumi.com/docs/guides/continuous-delivery/jenkins/), managed CI/CD platforms like Azure DevOps, and cloud native continuous delivery platforms like [Spinnaker](https://www.pulumi.com/docs/guides/continuous-delivery/spinnaker/) (see below for more details!). At the same time, we have heard feedback from Pulumi users looking to automate their delivery processes inside Kubernetes using a GitOps approach to delivery - both for Kubernetes workloads and for cloud infrastructure in any [Pulumi-supported Cloud Provider](https://www.pulumi.com/docs/intro/cloud-providers/).

Reviewer 2: ```suggestion
Pulumi integrates into [more than a dozen CI/CD and delivery systems]({{< relref “/docs/guides/continuous-delivery” >}}), automating the delivery of cloud infrastructure changes to AWS, Azure, GCP, Kubernetes, and more. This includes traditional CI tools like [Jenkins]({{< relref “/docs/guides/continuous-delivery/jenkins” >}}), managed CI/CD platforms like Azure DevOps, and cloud native continuous delivery platforms like [Spinnaker]({{< relref “/docs/guides/continuous-delivery/spinnaker” >}}) (see below for more details!). At the same time, we have heard feedback from Pulumi users looking to automate their delivery processes inside Kubernetes using a GitOps approach to delivery - both for Kubernetes workloads and for cloud infrastructure in any [Pulumi-supported Cloud Provider]({{< relref “/docs/intro/cloud-providers” >}}).
```

> We’re excited to announce the new [pulumi-kubernetes-operator](https://github.com/pulumi/pulumi-kubernetes-operator) open source project which provides a new Kubernetes-native option for infrastructure delivery. The Pulumi Kubernetes Operator allows you to specify the cloud infrastructure you want to deploy using a new `Stack` CustomResource in your Kubernetes cluster. The operator continuously drives the state of your cloud resources to align with the specified Stack. Stack resources can point at any Pulumi program in any Git repository, and use that program to define the desired state of the infrastructure. These stacks provide the full flexibility of Pulumi, including the ability to enforce policy using CrossGuard policy-as-code coupled with the benefits of the declarative Kubernetes resource model. You can create these Stack resources in normal Kubernetes YAML, or else choose to use Pulumi to manage the Kubernetes resources as well!

Reviewer 2: ```suggestion
We’re excited to announce the new [pulumi-kubernetes-operator](https://github.com/pulumi/pulumi-kubernetes-operator) open source project which provides a new Kubernetes-native option for infrastructure delivery. The Pulumi Kubernetes Operator allows you to specify the cloud infrastructure to deploy using a new `Stack` CustomResource in your Kubernetes cluster. The operator continuously drives the state of your cloud resources to align with the specified Stack. Stack resources can point at any Pulumi program in any Git repository, and use that program to define the desired state of the infrastructure. These stacks provide the full flexibility of Pulumi, including the ability to enforce policy using CrossGuard policy-as-code coupled with the benefits of the declarative Kubernetes resource model. You can create these Stack resources in normal Kubernetes YAML, or use Pulumi to manage the Kubernetes resources as well!
```

![Pulumi Kubernetes Operator](pulumi-kubernetes-operator.png)

> Creating a Stack is simple: provide the stack name, project repo, and a commit (support for branches coming soon!), and the operator starts driving toward the specified state. Stacks can be used directly, or as part of higher-level abstractions to manage your infrastructure and applications. The [Pulumi Console](https://www.pulumi.com/docs/intro/console/) augments these capabilities by providing visibility into the updates, resources under management, and policy enforcement of your stacks.

Reviewer 2: ```suggestion
Creating a Stack is simple: provide the stack name, project repo, and a commit (support for branches coming soon!), and the operator starts driving toward the specified state. Stacks can be used directly, or as part of higher-level abstractions to manage your infrastructure and applications. The [Pulumi Console]({{< relref “/docs/intro/console” >}}) augments these capabilities by providing visibility into the updates, resources under management, and policy enforcement of your stacks.
```

```yaml
apiVersion: pulumi.com/v1alpha1
kind: Stack
metadata:
  name: s3-bucket-stack
spec:
  stack: acmecorp/s3-op-project/dev
  projectRepo: https://github.com/joeduffy/test-s3-op-project
  commit: cc5442870f1195216d6bc340c14f8ae7d28cf3e2
  config:
    aws:region: us-east-2
  initOnCreate: true
  accessTokenSecret: pulumi-api-secret
  envSecrets:
    - pulumi-aws-secrets
```

## Pulumi Spinnaker Plugin

> Many teams using Spinnaker have asked about being able to automate the delivery of their cloud infrastructure changes using Pulumi through the same great continuous delivery platform that Spinnaker offers for application deployments.  As launch partners for the new Spinnaker Plugin model announced in June, Pulumi and Spinnaker worked closely together to deliver great support for Pulumi inside the Spinnaker continuous deployment platform.  

Reviewer 2: ```suggestion
Many teams using Spinnaker have asked about automating the delivery of their cloud infrastructure changes using Pulumi through the same great continuous delivery platform that Spinnaker offers for application deployments.  As launch partners for the new Spinnaker Plugin model announced in June, Pulumi and Spinnaker worked closely together to deliver complete support for Pulumi inside the Spinnaker continuous deployment platform.  
```

> You can check out the Pulumi Spinnaker integration in this presentation from Spinnaker Live and in the [Spinnaker guide](https://www.pulumi.com/docs/guides/continuous-delivery/spinnaker/) in the Pulumi docs.

Reviewer 2: ```suggestion
You can check out the Pulumi Spinnaker integration in this presentation from Spinnaker Live and in the [Spinnaker guide]({{< relref “/docs/guides/continuous-delivery/spinnaker” >}}) in the Pulumi docs.
```

For example, the commonly used [cert-manager](https://cert-manager.io/docs/installation/kubernetes/) CRD extends Kubernetes with the ability to easily configure Certificate Authorities and request certificates. After generating API wrappers with `crd2pulumi`, users can interact with the Cert Manager API from code:

<!-- TODO: Replace with an IDE screenshot to highlight IDE tooling and/or lang chooser. -->

```ts
import * as certificates from "./certificates"

new certificates.v1beta1.Certificate("example-cert", {
    metadata: {
        name: "example-com",
    },
    spec: {
        secretName: "example-com-tls",
        duration: "2160h",
        renewBefore: "360h",
        commonName: "example.com",
        dnsNames: [
            "example.com",
            "www.example.com",
        ],
        issuerRef: {
            name: "ca-issuer",
            kind: "Issuer",
        }
    }
});
```

## OPA support in Pulumi CrossGuard Policy-as-Code

> Pulumi’s [CrossGuard](https://www.pulumi.com/docs/guides/crossguard/) policy-as-code framework enforces policy across all your modern cloud infrastructure and/or in Kubernetes - using the full expressiveness of rich programming languages. It is deeply integrated into the deployment process to ensure that policy violations are identified *before* infrastructure is deployed instead of minutes or hours after changes are made.

Reviewer 2: ```suggestion
Pulumi’s [CrossGuard]({{< relref “/docs/guides/crossguard” >}}) policy-as-code framework enforces policy across all your modern cloud infrastructure and/or in Kubernetes - using the full expressiveness of rich programming languages. It is deeply integrated into the deployment process to ensure that policy violations are identified *before* infrastructure is deployed instead of minutes or hours after changes are made.
```

OPA-based rules can be authored using Pulumi’s rich resource model (top pane) and enforced as part of deployments to any infrastructure (bottom pane).

![Pulumi OPA CrossGuard VSCode](opa-crossguard-vscode.png)

And as part of Pulumi’s support for organizational enforcement of policy, OPA-based policy can be enforced across all infrastructure deployments in your organization (including across different clouds and in Kubernetes clusters).

![Pulumi OPA CrossGuard Console](opa-crossguard-console.png)

<!-- TODO: Add Helm 3 and Kustomize section -->

## Conclusion

We’re incredibly excited about today’s new features and the opportunity to deepen Pulumi’s support for Kubernetes and the Cloud Native ecosystem. The Pulumi Kubernetes Operator unlocks exciting new approaches to cloud native infrastructure delivery. The `kube2pulumi` tool makes it easier than ever to migrate from cumbersome YAML to expressive code. Strongly-typed CRDs bring the richness of Pulumi to the entire Cloud Native ecosystem of extensions to the Kubernetes platform. And OPA support in Pulumi policy-as-code brings cloud native policy authoring with OPA into the Pulumi Modern Infrastructure-as-Code world.

> It’s never been a more exciting time to get started with Pulumi and Kubernetes for your next modern cloud project.  Check out the new features or the Pulumi Kubernetes [Getting Started](https://www.pulumi.com/docs/get-started/kubernetes/) today.  

Reviewer 2: ```suggestion
It’s never been a more exciting time to get started with Pulumi and Kubernetes for your next modern cloud project.  Check out the new features or the Pulumi Kubernetes [Getting Started]({{< relref “/docs/get-started/kubernetes/“ >}}) today.  
```

You can learn more about each of the new capabilities in their corresponding project documentation:

* [Pulumi Kubernetes Operator](https://github.com/pulumi/pulumi-kubernetes-operator)
* [Strongly-typed CRDs](https://cert-manager.io/docs/installation/kubernetes/)
* [kube2pulumi](https://github.com/pulumi/kube2pulumi)
* [OPA for Pulumi CrossGuard](https://github.com/pulumi/pulumi-policy-opa)

> * [Spinnaker Pulumi Plugin](https://www.pulumi.com/blog/unlocking-spinnaker-with-pulumi/)

Reviewer 2: ```suggestion
* [Spinnaker Pulumi Plugin]({{< relref “/blog/unlocking-spinnaker-with-pulumi/“ >}})
```

---
title: "New Kubernetes Superpowers for Cloud Native Infrastructure"
date: 2020-08-12
meta_desc: "Announcing new features that advance Pulumi's Kubernetes support for deployment automation, cloud native ecosystem integration, and easy Pulumi adoption."
meta_image: k8s-enhancements.png
authors:
    - luke-hoban
tags:
    - kubernetes
---

> Today, we’re excited to [announce](https://info.pulumi.com/kubernetes) new features that advance Pulumi’s Kubernetes support.  These capabilities have been developed while helping leading cloud native engineering teams like Snowflake, Lemonade, and Mercedes-Benz go into production with Kubernetes, and include new deployment automation options, improved ecosystem integrations, and tools to make it easier than ever to adopt Pulumi for new and existing projects.

Author: Yes - that's where we publish press releases.

> Today, we’re excited to [announce](https://info.pulumi.com/kubernetes) new features that advance Pulumi’s Kubernetes support.  These capabilities have been developed while helping leading cloud native engineering teams like Snowflake, Lemonade, and Mercedes-Benz go into production with Kubernetes, and include new deployment automation options, improved ecosystem integrations, and tools to make it easier than ever to adopt Pulumi for new and existing projects.

Author: (I think it's owned by Hubspot)

For example, the commonly used [cert-manager](https://cert-manager.io/docs/installation/kubernetes/) CRD extends Kubernetes with the ability to easily configure Certificate Authorities and request certificates. After generating API wrappers with `crd2pulumi`, users can interact with the Cert Manager API from code:

![IDE type checking](ide-hint.png)

> {{< chooser language "typescript,python,csharp,go" >}}

Reviewer 3: Nevermind, I see it in the latest preview.

Today’s release of [Open Policy Agent](https://www.openpolicyagent.org/) (OPA) support in Pulumi policy-as-code extends the existing support for TypeScript and Python to include OPA’s `.rego` language and runtime for declarative policy specification. As an emerging standard for cloud native policy, OPA is already used across a wide variety of cloud native technologies, and can now also be leveraged to author policies to run in Pulumi’s policy model across any cloud or Kubernetes.

OPA-based rules can be authored using Pulumi’s rich resource model (top pane) and enforced as part of deployments to any infrastructure (bottom pane).

![Pulumi OPA CrossGuard VSCode](opa-crossguard-vscode.png)

And as part of Pulumi’s support for organizational enforcement of policy, OPA-based policy can be enforced across all infrastructure deployments in your organization (including across different clouds and in Kubernetes clusters).

![Pulumi OPA CrossGuard Console](opa-crossguard-console.png)

> <!-- TODO: Add Helm 3 and Kustomize section -->

Reviewer 3: Did you mean to leave this?

Author: Added the section now.

Today’s release of [Open Policy Agent](https://www.openpolicyagent.org/) (OPA) support in Pulumi policy-as-code extends the existing support for TypeScript and Python to include OPA’s `.rego` language and runtime for declarative policy specification. As an emerging standard for cloud native policy, OPA is already used across a wide variety of cloud native technologies, and can now also be leveraged to author policies to run in Pulumi’s policy model across any cloud or Kubernetes.

OPA-based rules can be authored using Pulumi’s rich resource model (top pane) and enforced as part of deployments to any infrastructure (bottom pane).

![Pulumi OPA CrossGuard VSCode](opa-crossguard-vscode.png)

And as part of Pulumi’s support for organizational enforcement of policy, OPA-based policy can be enforced across all infrastructure deployments in your organization (including across different clouds and in Kubernetes clusters).

![Pulumi OPA CrossGuard Console](opa-crossguard-console.png)

## Helm 3 and Kustomize

> Pulumi's Kubernetes support has long included support for both defining Kubernetes resources directly in your Pulumi program, as well as using Pulumi to deploy Kubernetes resources define in existing YAML or Helm. With a wealth of existing YAML and Helm content out in the ecosystem, these features have been used heavily by cloud engineering teams to deploy critical Kubernetes platform services and application logic into their Kuberentes clusters.  

Reviewer: ```suggestion
Pulumi's Kubernetes support has long included support for both defining Kubernetes resources directly in your Pulumi program, as well as using Pulumi to deploy Kubernetes resources from existing YAML or Helm. With a wealth of existing YAML and Helm content out in the ecosystem, these features have been used heavily by cloud engineering teams to deploy critical Kubernetes platform services and application logic into their Kubernetes clusters.  
```

## Pulumi Kubernetes Operator

> Pulumi integrates into [more than a dozen CI/CD and delivery systems]({{< relref “/docs/guides/continuous-delivery” >}}), automating the delivery of cloud infrastructure changes to AWS, Azure, GCP, Kubernetes, and more. This includes traditional CI tools like [Jenkins]({{< relref “/docs/guides/continuous-delivery/jenkins” >}}), managed CI/CD platforms like Azure DevOps, and cloud native continuous delivery platforms like [Spinnaker]({{< relref “/docs/guides/continuous-delivery/spinnaker” >}}) (see below for more details!). At the same time, we have heard feedback from Pulumi users looking to automate their delivery processes inside Kubernetes using a GitOps approach to delivery - both for Kubernetes workloads and for cloud infrastructure in any [Pulumi-supported Cloud Provider]({{< relref “/docs/intro/cloud-providers” >}}).

Reviewer 2: ```suggestion
Pulumi integrates into [more than a dozen CI/CD and delivery systems]({{< relref "/docs/guides/continuous-delivery" >}}), automating the delivery of cloud infrastructure changes to AWS, Azure, GCP, Kubernetes, and more. This includes traditional CI tools like [Jenkins]({{< relref "/docs/guides/continuous-delivery/jenkins" >}}), managed CI/CD platforms like Azure DevOps, and cloud native continuous delivery platforms like [Spinnaker]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}) (see below for more details!). At the same time, we have heard feedback from Pulumi users looking to automate their delivery processes inside Kubernetes using a GitOps approach to delivery - both for Kubernetes workloads and for cloud infrastructure in any [Pulumi-supported Cloud Provider]({{< relref "/docs/intro/cloud-providers" >}}).
```

We’re excited to announce the new [pulumi-kubernetes-operator](https://github.com/pulumi/pulumi-kubernetes-operator) open source project which provides a new Kubernetes-native option for infrastructure delivery. The Pulumi Kubernetes Operator allows you to specify the cloud infrastructure to deploy using a new `Stack` CustomResource in your Kubernetes cluster. The operator continuously drives the state of your cloud resources to align with the specified Stack. Stack resources can point at any Pulumi program in any Git repository, and use that program to define the desired state of the infrastructure. These stacks provide the full flexibility of Pulumi, including the ability to enforce policy using CrossGuard policy-as-code coupled with the benefits of the declarative Kubernetes resource model. You can create these Stack resources in normal Kubernetes YAML, or use Pulumi to manage the Kubernetes resources as well!

![Pulumi Kubernetes Operator](pulumi-kubernetes-operator.png)

> Creating a Stack is simple: provide the stack name, project repo, and a commit (support for branches coming soon!), and the operator starts driving toward the specified state. Stacks can be used directly, or as part of higher-level abstractions to manage your infrastructure and applications. The [Pulumi Console]({{< relref “/docs/intro/console” >}}) augments these capabilities by providing visibility into the updates, resources under management, and policy enforcement of your stacks.

Reviewer 2: ```suggestion
Creating a Stack is simple: provide the stack name, project repo, and a commit (support for branches coming soon!), and the operator starts driving toward the specified state. Stacks can be used directly, or as part of higher-level abstractions to manage your infrastructure and applications. The [Pulumi Console]({{< relref "/docs/intro/console" >}}) augments these capabilities by providing visibility into the updates, resources under management, and policy enforcement of your stacks.
```

## Pulumi Spinnaker Plugin

Many teams using Spinnaker have asked about automating the delivery of their cloud infrastructure changes using Pulumi through the same great continuous delivery platform that Spinnaker offers for application deployments.  As launch partners for the new Spinnaker Plugin model announced in June, Pulumi and Spinnaker worked closely together to deliver complete support for Pulumi inside the Spinnaker continuous deployment platform.  

> You can check out the Pulumi Spinnaker integration in this presentation from Spinnaker Live and in the [Spinnaker guide]({{< relref “/docs/guides/continuous-delivery/spinnaker” >}}) in the Pulumi docs.

Reviewer 2: ```suggestion
You can check out the Pulumi Spinnaker integration in this presentation from Spinnaker Live and in the [Spinnaker guide]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}) in the Pulumi docs.
```

## OPA support in Pulumi CrossGuard Policy-as-Code

> Pulumi’s [CrossGuard]({{< relref “/docs/guides/crossguard” >}}) policy-as-code framework enforces policy across all your modern cloud infrastructure and/or in Kubernetes - using the full expressiveness of rich programming languages. It is deeply integrated into the deployment process to ensure that policy violations are identified *before* infrastructure is deployed instead of minutes or hours after changes are made.

Reviewer 2: ```suggestion
Pulumi’s [CrossGuard]({{< relref "/docs/guides/crossguard" >}}) policy-as-code framework enforces policy across all your modern cloud infrastructure and/or in Kubernetes - using the full expressiveness of rich programming languages. It is deeply integrated into the deployment process to ensure that policy violations are identified *before* infrastructure is deployed instead of minutes or hours after changes are made.
```

We’re incredibly excited about today’s new features and the opportunity to deepen Pulumi’s support for Kubernetes and the Cloud Native ecosystem. The Pulumi Kubernetes Operator unlocks exciting new approaches to cloud native infrastructure delivery. The `kube2pulumi` tool makes it easier than ever to migrate from cumbersome YAML to expressive code. Strongly-typed CRDs bring the richness of Pulumi to the entire Cloud Native ecosystem of extensions to the Kubernetes platform. And OPA support in Pulumi policy-as-code brings cloud native policy authoring with OPA into the Pulumi Modern Infrastructure-as-Code world.

> It’s never been a more exciting time to get started with Pulumi and Kubernetes for your next modern cloud project.  Check out the new features or the Pulumi Kubernetes [Getting Started]({{< relref “/docs/get-started/kubernetes“ >}}) today.  

Reviewer 2: ```suggestion
It’s never been a more exciting time to get started with Pulumi and Kubernetes for your next modern cloud project.  Check out the new features or the Pulumi Kubernetes [Getting Started]({{< relref "/docs/get-started/kubernetes" >}}) today.  
```

You can learn more about each of the new capabilities in their corresponding project documentation:

* [Pulumi Kubernetes Operator](https://github.com/pulumi/pulumi-kubernetes-operator)
* [Strongly-typed CRDs](https://cert-manager.io/docs/installation/kubernetes/)
* [kube2pulumi](https://github.com/pulumi/kube2pulumi)
* [OPA for Pulumi CrossGuard](https://github.com/pulumi/pulumi-policy-opa)

> * [Spinnaker Pulumi Plugin]({{< relref “/blog/unlocking-spinnaker-with-pulumi“ >}})

Reviewer 2: ```suggestion
* [Spinnaker Pulumi Plugin]({{< relref "/blog/unlocking-spinnaker-with-pulumi" >}})
```
